scipy/optimize/_numdiff.py

```diff
@@ -491,32 +491,43 @@ def _dense_difference(fun, x0, f0, h, use_one_sided, method):
     h_vecs = np.diag(h)
 
     for i in range(h.size):
+        df = np.zeros_like(f0)
```

Comment:
It looks like `df` will be all zeros if any of the `dx` are zero. Don't we want the elements of `df` corresponding with nonzero `dx` to be evaluated as usual?

Reply:
`dx` can either be scalar or `(N,)` length array.

If *all* of `dx` are zero then there don't need to be any evaluations of `fun(x)`; `df` should be 0 (or actually `np.zeros_like(f0)`), because the step is 0. This has the effect of reducing overall function evaluations.

If there are *any* non-zero `dx` then `fun(x)` is evaluated to calculate `df` as usual.

Reply:
<face-palm> i was tired.

Reply:
I wonder if the tests are hitting the array `dx` case. I think you will get a:
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

How about `np.any(dx)`?

Reply:
for future reference `dx` is an array of size > 1 when evaluating a hessian.